File: apps/api/src/db/migrate.ts
import fs from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { pool } from './pool.js';
import { logger } from '../utils/logger.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const migrationsDir = path.resolve(__dirname, '../../../../infra/migrations');

async function run() {
  const files = (await fs.readdir(migrationsDir)).filter((f) => f.endsWith('.sql')).sort();
  for (const file of files) {
    const sql = await fs.readFile(path.join(migrationsDir, file), 'utf-8');
    logger.info({ file }, 'Running migration');
    await pool.query(sql);
  }
  logger.info('Migrations completed');
  await pool.end();
}

run().catch(async (error) => {
  logger.error({ error }, 'Migration failed');
  await pool.end();
  process.exit(1);
});

File: apps/api-python/scripts/migrate.py
import logging
import os
import sys
from pathlib import Path

import psycopg2
from dotenv import load_dotenv

# Setup logging
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger(__name__)

load_dotenv()

DATABASE_URL = os.getenv("DATABASE_URL", "")
if not DATABASE_URL:
    logger.error("DATABASE_URL is required")
    sys.exit(1)

migrations_dir = Path(__file__).resolve().parents[3] / "infra" / "migrations"

try:
    logger.info(
        f"Connecting to database: {DATABASE_URL.split('@')[1] if '@' in DATABASE_URL else 'unknown'}"
    )
    with psycopg2.connect(DATABASE_URL) as conn:
        logger.info("Database connection established")
        with conn.cursor() as cur:
            migration_files = sorted(migrations_dir.glob("*.sql"))
            logger.info(f"Found {len(migration_files)} migration files")

            for migration in migration_files:
                logger.info(f"Executing migration: {migration.name}")
                sql = migration.read_text(encoding="utf-8")
                cur.execute(sql)
                conn.commit()

            logger.info("All migrations completed successfully")
except Exception as e:
    logger.error(f"Migration failed: {str(e)}", exc_info=True)
    sys.exit(1)
